[#66464] Fix oauth_user_token strategy.

ba1ash opened the pull request:
# Ticket
<!-- Provide the link to respective work package -->

<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->
https://community.openproject.org/wp/66464
# What are you trying to accomplish?
Fix issue with access token not being used right after obtaining. 
<!-- Provide a description of the changes. -->

# How?
Explicitly set new access_token to Authorization header.

# Merge checklist

- [x] Added/updated tests

ba1ash committed:
[#66464] Fix oauth_user_token strategy.

https://community.openproject.org/work_packages/66464

ba1ash committed:
Merge branch 'release/16.3' into bug/66464-user-token-to-nextcloud-is-invalid-after-openproject-deploy

ba1ash committed:
[#66464] Add a spec.

The spec ensures that newly generated access token is used right after being obtained in the
same thread.

NobodysNightmare commented on modules/storages/app/common/storages/adapters/authentication_strategies/oauth_user_token.rb:
```diff
@@ -75,8 +75,8 @@ def refresh_and_retry(storage, token, options, &)
           config = storage.oauth_configuration.to_httpx_oauth_config.to_h
           refreshed_session = refresh_token(config, options, token).value_or { return Failure(it) }
```
Sidenote (not really in scope of this PR, but I notice it again): I find this code hard to read.

Specifically, I am always surprised by the fact that execution can stop in this line in case of an error, because the `return Failure(...)` is not a JavaScript lambda where we return from the lambda, but it's a Ruby `return` that returns from the surrounding method (i.e. `refresh_and_retry`) .

NobodysNightmare commented on modules/storages/app/common/storages/adapters/authentication_strategies/oauth_user_token.rb:
```diff
@@ -75,8 +75,8 @@ def refresh_and_retry(storage, token, options, &)
-
-          yield refreshed_session
+          options = options.deep_merge(headers: { "Authorization" => "Bearer #{token.access_token}" })
+          yield OpenProject.httpx.with(options)
```
Is that the right place to fix this issue?

`refresh_token` already returns an HTTPX session and I see no reason why it should intend to return an unusable session that has to be fixed here.

Shouldn't (from the intention of the implementation) `refreshed_session` be usable here and we should add the `options.deep_merge` to `refresh_token`?

Alternatively `refresh_token` could just return the updated token and we take care of the rest here.

ba1ash replied:
I agree. Let me try to refactor a bit.

NobodysNightmare commented:
I just noticed that my comments were pending as a review -.-

I intended to comment immediately.

ba1ash committed:
[#66464] Try to simplify code a bit.

- update! method is used for token updates. If something goes off with saving we must raise an error and not continue
operation.
- for refreshing token httpx does not need Authorization header with access_token.

NobodysNightmare reviewed (approved):
The refactoring was bigger than I intended, but I like it a lot. Especially our... _creative_... use of HTTPX to obtain a token (that we then store) feels much more encapsulated right now. I find it easier to follow what's happening.

Bonus points for replacing an unchecked `update` with `update!` ;-)
